how_to_eurec4a/cloudmasks.md

```diff
@@ -4,136 +4,275 @@ jupytext:
+import numpy as np
+import xarray as xr
+import matplotlib as mpl
+mpl.rcParams['font.size'] = 12
+
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+from matplotlib.colors import LogNorm
+```
+
+We generally don't support ignoring warnings, however, we do so in this notebook to suppress warnings originating from zero division. Unfortunately, we coudn't find another way to handle the warnings. If you have an idea, please make a pull request and change it :)
+
+```{code-cell} ipython3
+import warnings
+warnings.filterwarnings("ignore")
+```
+
+### Cloud fraction functions
+We define some utility functions to extract (circle) cloud cover estimates from different datasets assuming that the datasets follow a certain flag meaning convention.  
+
+In particular, we define a **minimum cloud fraction** based on the cloud mask flag `most_likely_cloudy` and a **maximum cloud fraction** combining the flags `most_likely_cloudy` and `probably_cloudy`. The cloud mask datasets slightly vary in their flag definition for cloud free conditions and their handling of unvalid measurements which is taken care of by the following functions.
+
+```{code-cell} ipython3
+def _isvalid(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return ~(np.isnan(da) | ("unknown" in meanings and da==meanings["unknown"]))
+
+def _cloudy_max(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return (da==meanings["most_likely_cloudy"]) | (da==meanings["probably_cloudy"])
+
+def _cloudy_min(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return da==meanings["most_likely_cloudy"]
+
+def cfmin(ds):
+    sumdims = [d for d in ds.cloud_mask.dims if d != "time"]
+    return (_cloudy_min(ds.cloud_mask).sum(dim=sumdims)
+            / _isvalid(ds.cloud_mask).sum(dim=sumdims))
+
+def cfmax(ds):
+    sumdims = [d for d in ds.cloud_mask.dims if d != "time"]
+    return (_cloudy_max(ds.cloud_mask).sum(dim=sumdims)
+            / _isvalid(ds.cloud_mask).sum(dim=sumdims))
+
+def correct_VELOX(ds):
+    return ds.assign(CF_min=ds.CF_min.where((ds.CF_min >=0 ) & (ds.CF_min <= 1)),
+                     CF_max=ds.CF_max.where((ds.CF_max >=0 ) & (ds.CF_max <= 1)))
+
+def ensure_cfminmax(ds):
+    if "CF_min" not in ds:
+        ds = ds.assign(CF_min=cfmin)
+    if "CF_max" not in ds:
+        ds = ds.assign(CF_max=cfmax)
+    return correct_VELOX(ds)
+
+def midpoint(a, b):
+    return a + (b - a) / 2
+
+def cf_circles(ds):
+    return xr.concat(
+        [
+            xr.Dataset({
+                "CF_max": ds.sel(time=slice(i["start"], i["end"])).CF_max.mean(dim="time"),
+                "CF_min": ds.sel(time=slice(i["start"], i["end"])).CF_min.mean(dim="time"),
+                "time": xr.DataArray(midpoint(i["start"], i["end"]), dims=()),
+                "segment_id": xr.DataArray(i["segment_id"], dims=())
+            })
+            for i in segments.values()
+        ], dim="time")
+
```

Comment:
To ease the understanding which preprocessing steps have been taken on the dataset to generate which plot, I would move the functions that are only used for a specific analysis/plot to that section and leave here only the functions used generally/for all analysis in the notebook, that is moving the `midpoint` and `cf_circles` to the statistical comparison section.

Reply:
That's a good point. I moved the functions you mentioned to the respective section.